File: backend/app/services/orchestrator.py
from time import perf_counter
from uuid import uuid4

from app.core.config import get_settings
from app.models.schemas import ChatRequest, ChatResponse, Source
from app.services.critic import CriticService
from app.services.guardrails import GuardrailService
from app.services.llm_provider import LLMProvider
from app.services.retriever import VectorRetriever
from app.services.router import RouterService
from app.services.telemetry import TelemetryService


class ChatOrchestrator:
    def __init__(
        self,
        router: RouterService,
        retriever: VectorRetriever,
        provider: LLMProvider,
        critic: CriticService,
        guardrails: GuardrailService,
        telemetry: TelemetryService,
    ) -> None:
        self.router = router
        self.retriever = retriever
        self.provider = provider
        self.critic = critic
        self.guardrails = guardrails
        self.telemetry = telemetry
        self.settings = get_settings()

    async def run(self, request: ChatRequest) -> ChatResponse:
        trace_id = str(uuid4())
        total_start = perf_counter()

        guard_start = perf_counter()
        guardrail = self.guardrails.check(request.question)
        self.telemetry.log(
            trace_id,
            "guardrails",
            {"blocked": guardrail.blocked, "reason": guardrail.reason},
            elapsed_ms=(perf_counter() - guard_start) * 1000,
        )
        if guardrail.blocked:
            response = ChatResponse(
                answer=guardrail.response or "Request blocked by guardrails.",
                sources=[],
                model=self.settings.gemini_model,
                route="reject",
                confidence=1.0,
                trace_id=trace_id,
                critic_passed=True,
            )
            self.telemetry.log(
                trace_id,
                "final",
                {"route": response.route, "blocked": True},
                elapsed_ms=(perf_counter() - total_start) * 1000,
            )
            return response

        route_start = perf_counter()
        decision = self.router.decide(request.question)
        self.telemetry.log(
            trace_id,
            "route",
            {"route": decision.route, "confidence": decision.confidence, "reason": decision.reason},
            elapsed_ms=(perf_counter() - route_start) * 1000,
        )

        if decision.route == "reject":
            response = ChatResponse(
                answer=(
                    "I can help with GitLab Handbook and Direction topics only. "
                    "Please ask a question related to GitLab processes, strategy, teams, or documentation."
                ),
                sources=[],
                model=self.settings.gemini_model,
                route=decision.route,
                confidence=decision.confidence,
                trace_id=trace_id,
                critic_passed=True,
            )
            self.telemetry.log(
                trace_id,
                "final",
                {"route": response.route, "blocked": True},
                elapsed_ms=(perf_counter() - total_start) * 1000,
            )
            return response

        if decision.route == "clarify":
            response = ChatResponse(
                answer=(
                    "Could you add a bit more detail so I can retrieve the right handbook or direction context? "
                    "For example, include the team, process, or objective you mean."
                ),
                sources=[],
                model=self.settings.gemini_model,
                route=decision.route,
                confidence=decision.confidence,
                trace_id=trace_id,
                critic_passed=True,
            )
            self.telemetry.log(
                trace_id,
                "final",
                {"route": response.route, "blocked": False},
                elapsed_ms=(perf_counter() - total_start) * 1000,
            )
            return response

        retrieval_mode = decision.route if decision.route in {"vector", "graph", "hybrid"} else "hybrid"
        retrieve_start = perf_counter()
        sources = await self.retriever.search_with_mode(
            request.question,
            top_k=self.settings.max_context_chunks,
            mode=retrieval_mode,
        )
        self.telemetry.log(
            trace_id,
            "retrieve",
            {"mode": retrieval_mode, "source_count": len(sources)},
            elapsed_ms=(perf_counter() - retrieve_start) * 1000,
        )

        generate_start = perf_counter()
        answer = await self.provider.generate(request.question, sources, request.history)
        self.telemetry.log(
            trace_id,
            "generate",
            {"answer_chars": len(answer)},
            elapsed_ms=(perf_counter() - generate_start) * 1000,
        )

        critic_start = perf_counter()
        critic_result = self.critic.evaluate(answer, sources)
        attempts = 0
        while (not critic_result.passed) and attempts < self.settings.max_regeneration_attempts:
            attempts += 1
            retry_question = (
                "Answer strictly from the provided context and avoid unsupported claims. "
                f"Original question: {request.question}"
            )
            answer = await self.provider.generate(retry_question, sources, request.history)
            critic_result = self.critic.evaluate(answer, sources)
        self.telemetry.log(
            trace_id,
            "critic",
            {"passed": critic_result.passed, "reason": critic_result.reason, "attempts": attempts},
            elapsed_ms=(perf_counter() - critic_start) * 1000,
        )

        # Generate follow-up suggestions in parallel with finalising the response
        followups = await self.provider.generate_followups(request.question, answer)

        response = ChatResponse(
            answer=answer,
            sources=sources,
            model=self.settings.gemini_model,
            route=decision.route,
            confidence=decision.confidence,
            trace_id=trace_id,
            critic_passed=critic_result.passed,
            followups=followups,
        )
        self.telemetry.log(
            trace_id,
            "final",
            {"route": response.route, "critic_passed": response.critic_passed},
            elapsed_ms=(perf_counter() - total_start) * 1000,
        )
        return response

File: nextapp/lib/orchestrator.js
import { checkGuardrails } from './guardrails.js';
import { routeQuery } from './router.js';
import { embedText, generateAnswer, generateFollowups } from './gemini.js';
import { searchVector } from './retriever.js';
import { evaluateCritic } from './critic.js';
import { v4 as uuidv4 } from 'uuid';

const TOP_K = 4;

/**
 * Run the full RAG pipeline and return a response object.
 * @param {string} question
 * @param {Array<{role: string, content: string}>} history
 * @returns {Promise<{answer, sources, route, traceId, criticPassed, followups, model}>}
 */
export async function orchestrate(question, history = []) {
  const traceId = uuidv4();
  const model = 'gemini-2.0-flash';

  // 1. Guardrails
  const guard = checkGuardrails(question, history);
  if (guard.blocked) {
    return {
      answer: guard.response || 'Request blocked by guardrails.',
      sources: [],
      route: 'reject',
      traceId,
      criticPassed: true,
      followups: [],
      model,
    };
  }

  // 2. Route
  const decision = routeQuery(question, history);

  if (decision.route === 'reject') {
    return {
      answer:
        'I can help with GitLab Handbook and Direction topics only. ' +
        'Please ask a question related to GitLab processes, strategy, teams, or documentation.',
      sources: [],
      route: decision.route,
      traceId,
      criticPassed: true,
      followups: [],
      model,
    };
  }

  if (decision.route === 'clarify') {
    return {
      answer:
        'Could you add a bit more detail so I can retrieve the right handbook or direction context? ' +
        'For example, include the team, process, or objective you mean.',
      sources: [],
      route: decision.route,
      traceId,
      criticPassed: true,
      followups: [],
      model,
    };
  }

  // 3. Build retrieval query
  let retrievalQuery = question;
  if (question.split(/\s+/).length <= 4 && history.length > 0) {
    const priorUserMsgs = history.filter((h) => h.role === 'user').map((h) => h.content);
    if (priorUserMsgs.length > 0) {
      retrievalQuery = `${priorUserMsgs[priorUserMsgs.length - 1]} ${question}`.trim();
    }
  }

  // 4. Retrieve sources via vector search
  let sources = [];
  try {
    const queryEmbedding = await embedText(retrievalQuery);
    sources = searchVector(queryEmbedding, TOP_K);
  } catch (err) {
    console.error('[orchestrator] Retrieval failed:', err.message);
    // Continue with empty sources (will likely fail critic but still respond)
  }

  // 5. Generate answer
  let answer = await generateAnswer(question, sources, history);

  // 6. Critic check with one retry
  let criticResult = evaluateCritic(answer, sources);
  if (!criticResult.passed) {
    const retryQuestion =
      'Answer strictly from the provided context and avoid unsupported claims. ' +
      `Original question: ${question}`;
    answer = await generateAnswer(retryQuestion, sources, history);
    criticResult = evaluateCritic(answer, sources);
  }

  // 7. Generate follow-ups
  let followups = [];
  try {
    followups = await generateFollowups(question, answer);
  } catch (err) {
    console.error('[orchestrator] Follow-ups failed:', err.message);
  }

  return {
    answer,
    sources,
    route: decision.route,
    traceId,
    criticPassed: criticResult.passed,
    followups,
    model,
  };
}
